### PR TITLE
Server-side validation MVP

### PR DESCRIPTION
**Pull request by landrito**
This is the minimum viable product for server side validation.

### CODE REVIEW

**Comment by landrito** [comment]
Disregard the changes to remove pagesize override which will be added in #9 and the changes to add the compile protos script which will be added in #8. I will rebase after those are merged.

File: server/session.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+	DeleteTest(name string) error
+	AddAnswers(name string, answers []string) error
+	TestAnswers(name string, answers []string) error
+}
+
+var instance Session
+var once sync.Once
+
+func GetSessionSingleton() Session {
+	once.Do(func() {
```
Comment (pongad) [inline server/session.go]: No need to use `once.Do`. Just
```
var Instance Session = &sessionImpl{...}
```
at top level

File: server/session.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+			tests: map[string]Test{},
+		}
+	})
+	return instance
+}
+
+type sessionImpl struct {
+	name string
+
+	mu    *sync.Mutex
```
Comment (pongad) [inline server/session.go]: mutex doesn't have to be a pointer.

File: server/session.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+	numOptional := 0
+	numOptionalSkipped := 0
+	numOptionalFailed := 0
+
+	errors := []*pb.ReportSessionResponse_Issue{}
+	warnings := []*pb.ReportSessionResponse_Issue{}
+
+	for _, test := range s.tests {
+		expectationLevel := test.GetExpectationLevel()
```
Comment (pongad) [inline server/session.go]: I think `expLvl` is a better name, but up to you :P

File: server/session.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	mu    *sync.Mutex
+	tests map[string]Test
+}
+
+func (s *sessionImpl) GetName() string {
+	return s.name
+}
+
+func (s *sessionImpl) GetReport() *pb.ReportSessionResponse {
+	numRequired := 0
```
Comment (pongad) [inline server/session.go]: I'd wrap it like this:
```
var (
numRequired int
numRequiredSkipped int
//...
errors []*pb.ResportSessionResponse_Issue // no need to init array, append already knows what to do
)
```

Also consider creating a type `type result struct {test, skipped, failed int}`

File: server/session.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+					numOptionalSkipped = numOptionalSkipped + 1
+				} else {
+					numOptionalFailed = numOptionalFailed + 1
+				}
+			}
+		}
+	}
+
+	var state pb.ReportSessionResponse_State
+	if numRequired == 0 || (numRequiredFailed == 0 && numRequiredSkipped == 0) {
```
Comment (pongad) [inline server/session.go]: shouldn't `numRequiredFailed == 0 && numRequiredSkipped ==0` be sufficient?

File: server/session.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	numOptionalFailed := 0
+
+	errors := []*pb.ReportSessionResponse_Issue{}
+	warnings := []*pb.ReportSessionResponse_Issue{}
+
+	for _, test := range s.tests {
+		expectationLevel := test.GetExpectationLevel()
+		issue := test.GetIssue()
+
+		if expectationLevel == pb.Test_REQUIRED {
```
Comment (pongad) [inline server/session.go]: If you create the `result` type, you can do
```
var r *result
switch expectationLevel {
case pb.Test_REQUIRED:
  r = &resultRequired
// ...
}

r.test++
if issue != nil {
  r.issues = append(r.issues, issue)
  if issue.Type == SKIPPED {
    r.skipped++
  } else {
    r.failed++
  }
}
```

Using pointer is also OK without the result type, but it's more awkward.

File: server/testing_service_test.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
+// Copyright 2018 Google LLC
```
Comment (pongad) [inline server/testing_service_test.go]: I won't stop you but I don't think these tests are actually necessary. `testingServerImpl` essentially forwards things to session.

File: server/testing_service.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+	err := s.session.DeleteTest(in.Name)
+	if err != nil {
+		return nil, err
+	}
+	return &empty.Empty{}, nil
+}
+
+func (s *testingServerImpl) RegisterTest(ctx context.Context, in *pb.RegisterTestRequest) (*empty.Empty, error) {
+	name := in.Name
+	name = strings.TrimPrefix(name, "/sessions/-")
```
Comment (pongad) [inline server/testing_service.go]: Do we want to error if the name doesn't start with the prefix?

Comment (landrito) [reply server/testing_service.go]: Eventually. I think what is present is fine for now until I have the multiple sessions story. For now we can simply defer to the session to do the error handling.

I'll make a comment in code.

File: server/testing_service_test.go (hunk 7 above)
Comment (landrito) [reply server/testing_service_test.go]: Agreed. Admittedly, I just wanted to keep 100% test coverage.

File: server/session.go (hunk 5 above)
Comment (landrito) [reply server/session.go]: Ah yes that's correct. Will change.

File: server/session.go (hunk 6 above)
Comment (landrito) [reply server/session.go]: Ah this is much better thanks! Done.

File: server/session.go (hunk 5 above)
Comment (landrito) [reply server/session.go]: Done.

File: server/session.go (hunk 4 above)
Comment (landrito) [reply server/session.go]: Done.

File: server/session.go (hunk 3 above)
Comment (landrito) [reply server/session.go]: Done!

File: server/session.go (hunk 2 above)
Comment (landrito) [reply server/session.go]: Done.

File: server/session.go (hunk 1 above)
Comment (landrito) [reply server/session.go]: Done.

**Pull request merged**